napari/_qt/menus/_tests/test_plugins_menu.py

```diff
@@ -21,19 +52,88 @@ def widget1():
     assert len(viewer.window._dock_widgets) == 0
-    plugin_action_menu.actions()[0].trigger()
+    assert 'tmp_plugin:Widget 1' in app.commands
+    # Trigger the action, opening the first widget: `Widget 1`
+    app.commands.execute_command('tmp_plugin:Widget 1')
     assert len(viewer.window._dock_widgets) == 1
-    assert (
-        next(iter(viewer.window._dock_widgets.data)) == 'Widget 1 (tmp_plugin)'
+    assert 'Widget 1 (Temp Plugin)' in viewer.window._dock_widgets
+
+
+def test_plugin_menu_plugin_state_change(
+    make_napari_viewer,
+    tmp_plugin: DynamicPlugin,
+):
+    """Check plugin menu items correct after a plugin changes state."""
+    app = get_app()
+    pm = tmp_plugin.plugin_manager
+
+    # Register plugin q actions
+    init_qactions()
+    # Check only `Q_PLUGINS_ACTIONS` in plugin menu before any plugins registered
+    plugins_menu = app.menus.get_menu(MenuId.MENUBAR_PLUGINS)
+    assert len(plugins_menu) == len(_plugins.Q_PLUGINS_ACTIONS)
+
+    @tmp_plugin.contribute.widget(display_name='Widget 1')
+    def widget1():
+        return DummyWidget()
+
+    @tmp_plugin.contribute.widget(display_name='Widget 2')
+    def widget2():
+        return DummyWidget()
+
+    # Configures `app`, registers actions and initializes plugins
+    make_napari_viewer()
+
+    plugins_menu = app.menus.get_menu(MenuId.MENUBAR_PLUGINS)
+    assert len(plugins_menu) == len(_plugins.Q_PLUGINS_ACTIONS) + 1
+    assert isinstance(plugins_menu[-1], SubmenuItem)
+    assert plugins_menu[-1].title == tmp_plugin.display_name
+    plugin_submenu = app.menus.get_menu(MenuId.MENUBAR_PLUGINS + '/tmp_plugin')
+    assert len(plugin_submenu) == 2
+    assert isinstance(plugin_submenu[0], MenuItem)
+    assert plugin_submenu[0].command.title == 'Widget 1'
+    assert 'tmp_plugin:Widget 1' in app.commands
+
+    # Disable plugin
+    pm.disable(tmp_plugin.name)
+    with pytest.raises(KeyError):
+        app.menus.get_menu(MenuId.MENUBAR_PLUGINS + '/tmp_plugin')
+    assert 'tmp_plugin:Widget 1' not in app.commands
+
+    # Enable plugin
+    pm.enable(tmp_plugin.name)
+    samples_sub_menu = app.menus.get_menu(
+        MenuId.MENUBAR_PLUGINS + '/tmp_plugin'
     )
+    assert len(samples_sub_menu) == 2
+    assert 'tmp_plugin:Widget 1' in app.commands
+
+
+def test_plugin_widget_checked(make_napari_viewer, tmp_plugin: DynamicPlugin):
+    """Check widget toggling updates check mark correctly."""
+
+    @tmp_plugin.contribute.widget(display_name='Widget')
+    def widget():
+        return DummyWidget()
+
+    app = get_app()
+    viewer = make_napari_viewer()
+
+    assert 'tmp_plugin:Widget' in app.commands
+    widget_action = viewer.window.plugins_menu.findAction('tmp_plugin:Widget')
+    # Trigger the action, opening the widget
+    assert not widget_action.isChecked()
+    widget_action.trigger()
+    assert widget_action.isChecked()
+    assert 'Widget (Temp Plugin)' in viewer.window._dock_widgets
```

Comment:
would love here to hide the widget and/or destroy it and make sure the check mark is not shown? I've confirmed visually that it works but would be nice to have a test

Reply:
I'm afraid it may be like the `isVisible` check that does not seem to work in tests. The `QAction` does not seem to get updated with hide or destroy. I've tried updating the plugin menu from context, which does not work. Suggestions welcome.

Reply:
Ahh yeah ok. I think that's ok to leave as is then